### ISSUE TITLE
Feature: allow Zod validation to run for client-only contract implementations

### ISSUE THREAD

**Issue by br0p0p** [ts-rest/ts-rest#514]
Hello!

First of all, I'd just like to say this library is great and makes defining API contracts a breeze. Thanks for working on it!

It seems that the `zod` schemas are not executed in client-only implementations. This confused me as the docs led me to believe that defining `zod` schemas within contracts will cause request payloads to be validated, regardless of whether the contract is used on client or server.

The validation of input data is useful to catch bugs in downstream code. 

Furthermore, using `zod` for transforming input data is extremely useful, even if you don't want failed validation to stop a request from being sent.

Beyond that, including `zod` schemas in a contract also means any consumer of the contract incurs the bundle size cost of `zod`. It seems wasteful to have defined a contract using `zod` schemas and not be able to use the benefits of `zod`, even though it is included in the browser bundle.

---

For context, I work on a client-only team and we use this library to define a contract against a backend API server which is managed by another team. So, we don't have control over the server-side validation and would love to have the additional control of optionally validating our requests against this API, as well as having access to `zod` transformations.

I would be happy to work on this functionality if needed.

**Comment by Gabrola** [comment]
The seems to be an extremely specific need; I've never heard of needing to run request validation on the client before. In addition:
> even if you don't want failed validation to stop a request from being sent.

Zod will not output "transformed" data if validation fails. That's just something that you can't do.

Another issue is that transformations are usually intended to run only once on the server. If you run transformations on the client, the data might end up failing validation on the sever or being improperly transformed twice. Think of something like `z.number().transform(v => v * v)`. As unrealistic as this specific example may be, we cannot account for the use-cases where some might be using non-idempotent transforms.

In your case, if you must, you can just use a custom client and manually call Zod's `.parse()` before passing the data to `tsRestFetchApi`

**Comment by ivan-kleshnin** [comment]
Zodios has a client-side validation and I can tell from own experience – it's a very bad idea however good it might have seen initially:
- Zod is a forced dependency in the browser
- error handling becomes much more complex (BE and FE errors, obviously, behave differently)

**Comment by isaachinman** [comment]
> I've never heard of needing to run request validation on the client before

It's called... Form validation.

I was just diving into `ts-rest`, and I think I misunderstood the value prop. If we are defining contracts between BE and FE, it makes sense for validation to happen at both levels, as is best practice with any fullstack application.

Any reasonable FE is going to have to do input validation _on top of_ `ts-rest`, because this package does not expose runtime validation for non-server contexts.

In short: modern frontends _do_ perform validation, and most already use `zod`.

Very odd to omit this kind of functionality and seems to assume that all frontends are old-school form-submission applications relying solely on BE 400s.

**Comment by isaachinman** [comment]
I think we're mostly saying the same thing.

But your stance seems to be: duplicate your business logic and have two different schemas for BE and FE validation.

I would suggest you expose the runtime validation (optionally) in the FE clients. Yes, an endpoint signature and a FE form can differ in content. But quite a bit of the time, they are identical. At any rate, zod schemas can be joined or extended. Just my two cents 👍
